File: src/geometry/map_surface.f90
module map_surface
use math
use evaluate
implicit none

private
public :: surface_distance_map

contains

subroutine surface_distance_map(n, p, uk, m, q, vk, cpw, nu, nv, ugrid, vgrid, &
                                udist, vdist, uparam, vparam, xy)
    !> Build a distance map of the surface.
    
    !f2py intent(in) n, p, uk, q, m, vk, cpw, nu, nv, ugrid, vgrid
    !f2py intent(out) udist, vdist, uparam, vparam, xy
    !f2py depend(n, p) uk
    !f2py depend(m, q) vk
    !f2py depend(n, m) cpw
    !f2py depend(nu) ugrid
    !f2py depend(nv) vgrid
    !f2py depend(nu, nv) udist, vdist, xy
    
    ! Input
    integer, intent(in) :: n, p, m, q, nu, nv
    double precision, intent(in) :: uk(0:n + p + 1), vk(0:m + q + 1)
    double precision, intent(in) :: cpw(0:n, 0:m, 0:3)
    double precision, intent(in) :: ugrid(nu), vgrid(nv)
    
    ! Ouptut
    double precision, intent(out) :: udist(nu, nv), vdist(nu, nv)
    double precision, intent(out) :: uparam(nu * nv), vparam(nu * nv)
    double precision, intent(out) :: xy(nu * nv, 2)

    ! Working
    integer :: i, j, k
    double precision :: d, dv, du, du_0, du_1, dv_0, dv_1
    double precision :: p0(3), p1(3), dp(3)
    
    udist(:, :) = 0.0d0
    vdist(:, :) = 0.0d0
    uparam(:) = 0.0d0
    vparam(:) = 0.0d0
    xy(:, :) = 0.0d0
    
    ! Map u-direction.
    do j = 1, nv
        dv = vgrid(j)
        d = 0.0d0
        do i = 2, nu
            du_0 = ugrid(i - 1)
            du_1 = ugrid(i)
            call surface_point(n, p, uk, m, q, vk, cpw, du_0, dv, p0)
            call surface_point(n, p, uk, m, q, vk, cpw, du_1, dv, p1)
            dp = p1 - p0
            d = d + norm(dp)
            udist(i, j) = d
        end do
    end do
    
    ! Map v-direction.
    do i = 1, nu
        du = ugrid(i)
        d = 0.0d0
        do j = 2, nv
            dv_0 = vgrid(j - 1)
            dv_1 = vgrid(j)
            call surface_point(n, p, uk, m, q, vk, cpw, du, dv_0, p0)
            call surface_point(n, p, uk, m, q, vk, cpw, du, dv_1, p1)
            dp = p1 - p0
            d = d + norm(dp)
            vdist(i, j) = d
        end do
    end do
    
    ! Build data for inversion.
    k = 1
    do j = 1, nv
        do i = 1, nu
            xy(k, :) = (/ udist(i, j), vdist(i, j) /)
            uparam(k) = ugrid(i)
            vparam(k) = vgrid(j)
            k = k + 1
        end do
    end do

end subroutine surface_distance_map

end module map_surface

File: pynurbs/geometry/methods/map_surface.py
from numpy import float64, linspace, zeros, hstack, unique
from numpy.linalg import norm
from scipy.interpolate import LinearNDInterpolator, RectBivariateSpline


def surface_distance_map(surface, n=None, h=None):
    """
    Build a distance map of the
    """
    if n is None and h is None:
        return None, None, None, None

    # Use length of bounding box diagonal to estimate grid steps.
    if h is not None:
        bbox = surface.bbox
        diag = bbox.diagonal_length()
        ntemp = int(diag / h) + 1
        if ntemp > n:
            n = ntemp

    # Parameter domain.
    umin = surface.au
    umax = surface.bu
    vmin = surface.av
    vmax = surface.bv

    # Get knots of surfaces.
    uknots, vknots = surface.uknots, surface.vknots

    # Grid of uv values.
    ugrid = linspace(umin, umax, n)
    vgrid = linspace(vmin, vmax, n)

    # Combine and make unique.
    ugrid = hstack((ugrid, uknots))
    vgrid = hstack((vgrid, vknots))
    ugrid = unique(ugrid)
    vgrid = unique(vgrid)

    # Number of grids.
    nu = ugrid.size
    nv = vgrid.size

    # Surface evaluator.
    seval = surface.eval

    # Map u-direction.
    udist = zeros((nu, nv), dtype=float64)
    for j in range(0, nv):
        dv = vgrid[j]
        d = 0.
        for i in range(1, nu):
            du_0 = ugrid[i - 1]
            du_1 = ugrid[i]
            pi = seval(du_0, dv, domain='global', rtype='ndarray')
            pi1 = seval(du_1, dv, domain='global', rtype='ndarray')
            d += norm(pi1 - pi)
            udist[i, j] = d

    # Map v-direction.
    vdist = zeros((nu, nv), dtype=float64)
    for i in range(0, nu):
        du = ugrid[i]
        d = 0.
        for j in range(1, nv):
            dv_0 = vgrid[j - 1]
            dv_1 = vgrid[j]
            pi = seval(du, dv_0, domain='global', rtype='ndarray')
            pi1 = seval(du, dv_1, domain='global', rtype='ndarray')
            d += norm(pi1 - pi)
            vdist[i, j] = d

    # Build interpolation functions for parameters.
    xy = []
    uparam = []
    vparam = []
    for j in range(0, nv):
        for i in range(0, nu):
            xy.append([udist[i, j], vdist[i, j]])
            uparam.append(ugrid[i])
            vparam.append(vgrid[j])

    # Build interpolation functions for distance.
    interp_udist = RectBivariateSpline(ugrid, vgrid, udist, kx=1, ky=1)
    interp_vdist = RectBivariateSpline(ugrid, vgrid, vdist, kx=1, ky=1)
    interp_uparam = LinearNDInterpolator(xy, uparam, rescale=True)
    interp_vparam = LinearNDInterpolator(xy, vparam, rescale=True)

    return interp_udist, interp_vdist, interp_uparam, interp_vparam
